### ISSUE TITLE
Can't connect to docker container's exposed port with network's gateway ip from other container

### ISSUE THREAD

**Issue by illesguy** [docker/for-mac#5588]
- [x] I have tried with the latest version of Docker Desktop
  - [ ] I have tried disabling enabled experimental features
  - [ ] I have uploaded Diagnostics

### Expected behavior
When creating a docker container with an exposed port (not binding it), be able to reach the application running on it from another container in the same network using the network's gateway ip and the port the container was assigned on the host. `HostPort` in the docker container's `HostConfig`, `PortBindings` should be empty. This is the behaviour on Ubuntu.

### Actual behavior
When creating a docker container with an exposed port (not binding it), can reach the container from the host with localhost and the newly assigned port but not from another container with their network's gateway ip and the newly assigned port. `HostPort` in the docker container's `HostConfig, `PortBindings` is set to "0". This is the behaviour on Mac.

### Information

The problem is reproducible. Not sure if it's new, we just noticed it and the not sure whether it came with an update. Added steps to reproduce below. When the run a container and use the `-p <port>` option to map the container's port to a randomly assigned one on the host machine, we noticed that HostPort under HostConfig and PortBindings in the container's config is set to "0", on Ubuntu it is an empty string. If we were to specify the port we want to bind to it works fine `-p <host_port>:<container_port>`. In this case the `HostPort` under `HostConfig`/`PortBindings` would be the `host_port` we specified on both Mac and Ubuntu. We suspect this is causing the issue.

macOS version: 10.15.6
Docker Desktop Version: 3.3.1
Docker version: 20.10.5

Ubuntu host tested on: 18.04.5
Docker version: 20.10.6

### Reproduce

The following script would reproduce the issue, it would fail on Mac and succeed on Ubuntu:

```bash
  #!/usr/bin/env bash

echo "Starting PostgreSQL database in container"
CONTAINER=$(docker run -d -p 5432 -e POSTGRES_USER=test -e POSTGRES_PASSWORD=test -e POSTGRES_DB=test postgres:9.6)
echo "PostgreSQL database started in container: $CONTAINER"
sleep 5  # Making sure db will be responsive

PORT=$(docker port $CONTAINER | head -n 1 | rev | cut -d: -f1 | rev)
GATEWAY_IP=$(docker network inspect -f '{{range .IPAM.Config}}{{.Gateway}}{{end}}' bridge)
echo
echo "Gateway IP: $GATEWAY_IP, port: $PORT"

echo
echo "Querying database version from another container"
docker run postgres:9.6 psql "host=$GATEWAY_IP port=$PORT dbname=test user=test password=test" -c "select version()"

echo
echo "HostConfig PortBindings of container: $CONTAINER"
docker inspect -f '{{range $p, $conf := .HostConfig.PortBindings}} {{$p}} -> HostIp: {{(index $conf 0).HostIp}} HostPort: {{(index $conf 0).HostPort}} {{end}}' $CONTAINER

echo
echo "NetworkSettings Ports of container: $CONTAINER"
docker inspect -f '{{range $p, $conf := .NetworkSettings.Ports}} {{$p}} -> HostIp: {{(index $conf 0).HostIp}} HostPort: {{(index $conf 0).HostPort}} {{end}}' $CONTAINER

echo
echo "Stopping $CONTAINER"
docker stop $CONTAINER
```

**Comment by nk412** [comment]
+1 seeing the same issue

**Comment by djs55** [comment]
Thanks for your report.

For maximum compatibility/portability between systems I recommend
- using `docker run -p` to expose ports on the host (as you are doing)
- but to access one container from another on the same network use the internal DNS name and IP, for example:
```
% docker network create test  
% docker run --rm --net=test --name=container1 -it alpine sh
...
% docker run --rm --net=test --name=container2 -it alpine sh
/ # ping container1
PING container1 (172.18.0.2): 56 data bytes
64 bytes from 172.18.0.2: seq=0 ttl=64 time=0.596 ms
^C
--- container1 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 0.596/0.596/0.596 ms
```
The networking implementation on macOS is different to native Linux as there are two kernels with separate network stacks. On Linux `docker run -p` has the side-effect you observe, but on macOS the port has to be exposed differently, due to the different kernel.

**Comment by illesguy** [comment]
The issue we are having is within a Python library (https://github.com/testcontainers/testcontainers-python) that uses the approach I've described (network gateway ip + port) to communicate between containers it starts, so we can't change. For now we were able to fix this, by explicitly binding the ports we want to expose instead letting docker do it.

But if you're saying using the internal DNS and IP is recommended, maybe that is a ticket to open on the repository of the library we are using.

**Comment by ag-TJNII** [comment]
> On Linux `docker run -p` has the side-effect you observe, but on macOS the port has to be exposed differently, due to the different kernel.

So is this a 'Won't Fix', then?  This used to work fine on Docker for Mac.  I support Dockerized utilities in both Linux and Mac environments and use this behavior often.  I also tried hitting the listening container by it's IP when troubleshooting and that didn't work.  when did this behavior change?  I didn't see any notes about it in the changelogs.

I'm also unable to get back to a working state by uninstalling and installing an older version.  I'm suspicious this is due to a recent networking components update, and I noticed the uninstall does not ask for privileged access like the install does.  Does a uninstall also uninstall the networking components?

**Comment by goatherder** [comment]
I see the same behaviour with https://github.com/testcontainers/testcontainers-go

**Comment by djs55** [comment]
@rnorth sorry for the delay in replying. The observation about `0:8080` working but `8080` not working is interesting. I'll add a ticket to our backlog to investigate: it might be fixable as you suggest.

**Comment by djs55** [comment]
I've done a bit of investigating and I believe this problem is due to the fix for https://github.com/docker/for-win/issues/10008 . On Mac and Windows `dockerd` runs inside the VM, in a network namespace. When a random port is needed to expose a container (e.g. `docker run -p 80`) there was a bug where `dockerd` would choose a port which was free inside it's network namespace, but which wasn't free on the host. This was fixed by choosing a random port on both the host and on `127.0.0.1` in Linux, connecting them together, and rewriting the `docker ps` results over `/var/run/docker.sock` to show the correct value for the host:

![Desktop port forwarding (4)](https://user-images.githubusercontent.com/198586/136068651-69b2ead9-9a58-4b2d-9981-7c69bbb90ce1.png)

There was a bug in this bugfix, in the test for "is the user requesting a random port". The code checks for the port being "", but not the port being "0". So the reason that `-p 0:...` works at the moment is because it's avoiding this bugfix, by accident.

Considering the possibility of making `-p 80` and `-p 0:80` behave the same i.e. fixing the bug in the bugfix, I can see 2 further problems:

1. the bugfix binds internally on `127.0.0.1` inside the Linux network namespace, so the port isn't available on the gateway IP
2. even if the port were available on the gateway IP, it would be the internal port rather than the external one visible in `docker ps` on the host.

I could probably fix (1) by binding internally on `0.0.0.0` rather than `127.0.0.1`; hopefully this would be safe because the Linux network namespace with `dockerd` is almost empty. However this still leaves the problem of discovering the internal port number.

It is possible to bypass the Docker API proxy which is rewriting the port number:

* `docker run -v /var/run/docker.sock:/var/run/docker.sock docker ps`: would show the host port
* `docker run -v /var/run/docker.sock.raw:/var/run/docker.sock docker ps`: would show the internal port

I don't know how convenient this would be to change. It's a bit of an internal detail of Docker Desktop.

So in summary:
1. `-p 0:...` seems to work by accident rather than design 😱 
2. I can't guarantee anything but I think there's no imminent danger of it stopping working: I'll add a test case which checks that it continues to work and linking to this issue so can avoid breaking it if possible / give you advance warning
3. we need to think of a better fix for the original issue and this one

**Comment by gesellix** [comment]
> I'll add a test case which checks that it continues to work and linking to this issue so can avoid breaking it if possible / give you advance warning

❤️
